### PR TITLE
fix(grpo): penalize invalid tool call and malformed thinking

### LINKED ISSUE #2012: [super-pr] penalize toolcall in reasoning when it should be outside + invalid thinking tag
564c4855c6ee8aa5f69048905127190006625f50

### PR DESCRIPTION
**Pull request by macandro96**
# What does this PR do ?

Support to penalize textual tool calls and malformed thinking output. Currently, applicable only in legacy grpo and nemo-gym pathways.

Look at `Additional Information` below

# Issues
Closes https://github.com/NVIDIA-NeMo/RL/issues/2012

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

Adds GRPO support for penalizing malformed NeMo-Gym assistant outputs:

- flags invalid textual tool calls in NeMo-Gym rollouts, e.g. leaked `<tool_call>` / `<function_call>` tags
- flags malformed thinking output, e.g. `<think>` tags in final content or repeated thinking tags in reasoning summaries
- applies configured negative advantages to the flagged assistant-message token span
- adds `tis_clipped_fraction` to clipped PG loss metrics
- documents the new GRPO penalty knobs in the exemplar config
- wires optional NeMo-Gym detector pattern overrides through `run_grpo_nemo_gym.py`
- fails loudly for `data_plane.enabled=true`, where message-level advantage penalties are not supported yet

This combines behavior from:
- `2da2585984634b4f17a932a70c2b5d2854e217ff`
- `d6b27b5446818a2512785ea64397efd8a16fc10e`
- `b941004d2e0be87d67309763521c6b9e8d292ecd`
- `c21c87e640448734e2df4b3d3416abb4cd4080db`

### CODE REVIEW

File: nemo_rl/algorithms/loss/loss_functions.py [hunk 1]
```diff
@@ -457,7 +457,17 @@ def __call__(
...
         if self.truncated_importance_sampling_ratio is not None:
+            with torch.no_grad():
+                # Divide by global_valid_toks (not local mask.sum()) so that np.sum aggregation in grpo.py gives the correct global fraction
+                tis_clipped_fraction = (
+                    (
+                        actor_importance_weights_expanded
+                        > self.truncated_importance_sampling_ratio
+                    )
+                    & mask.bool()
+                ).sum().item() / max(global_valid_toks.item(), 1)
```
Comment (yuki-97) [inline nemo_rl/algorithms/loss/loss_functions.py]: looks `tis_clipped_fraction` is duplicated with `is_oob_ratio`, and they have different problems.

`is_oob_ratio`: the aggregate way is not correct. (not introduced by this PR)
`tis_clipped_fraction`: it seems only work for tis, not works for `icepop` and `seq-mask-tis`.

could you help to fix and unify to one metric?

File: examples/configs/grpo_math_1B.yaml [hunk 2]
```diff
@@ -38,6 +38,10 @@ grpo:
     target_min: 0.0
     target_max: 1.0
   seq_logprob_error_threshold: null
+  # Advantage value assigned to invalid tool call tokens (e.g. -5.0 to penalize). null disables.
+  invalid_tool_call_advantage: null
+  # Advantage value assigned to tokens with malformed <think>/</think> tags (e.g. -5.0). null disables.
+  malformed_thinking_advantage: null
```
Comment (yuki-97) [inline examples/configs/grpo_math_1B.yaml:46]: don't we need to also add these to `examples/nemo_gym/grpo_nanov3.yaml` and `examples/nemo_gym/grpo_workplace_assistant_nemotron_nano_v2_9b.yaml`?

File: nemo_rl/algorithms/grpo.py [hunk 3]
```diff
@@ -1074,6 +1080,64 @@ def add_grpo_token_loss_masks_and_generation_logprobs(
                 )
 
 
+def _apply_message_level_advantage_penalties(
```
Comment (yuki-97) [inline nemo_rl/algorithms/grpo.py:1119]: nit: does it make more sense to move it to `nemo_rl/algorithms/advantage_estimator.py` and remove the prefix `_`? wdyt?

File: nemo_rl/environments/nemo_gym.py [hunk 4]
```diff
@@ -221,6 +297,8 @@ def _postprocess_nemo_gym_to_nemo_rl_result(
                     "generation_logprobs": torch.tensor(
                         output_item_dict["generation_log_probs"]
                     ),
+                    "is_invalid_tool_call": is_invalid_tool_call,
+                    "has_malformed_thinking": has_malformed_thinking,
```
Comment (yuki-97) [inline nemo_rl/environments/nemo_gym.py:340]: looks like it's only supported for gym env and we won't get the two if we use native rollout?
if so, I think we should assert `invalid_tool_call_advantage` and `malformed_thinking_advantage` are None if not using gym env?

**Review by yuki-97 — COMMENTED** [review]
@macandro96 thanks for the effort, left some comments.

@yfw could you help to take a review as well when you have a chance?

File: nemo_rl/environments/nemo_gym.py (hunk 4 above)
Comment (macandro96) [reply nemo_rl/environments/nemo_gym.py:340]: @yfw mentioned that we plan to completely move to nemo-gym for rollouts in the future? hence we did not add for native ones.

Comment (yuki-97) [reply nemo_rl/environments/nemo_gym.py:340]: I think we will keep nemo-rl native async and nemo-gym path (removing native sync path) in the future, @terrykong can keep me honest.
so I think it should be good to just add an assert if not using nemo-gym path for now.

File: nemo_rl/algorithms/loss/loss_functions.py [hunk 5]
```diff
@@ -457,7 +457,17 @@ def __call__(
         #                  IS ratio ∉ [min, max]; retained sequences keep
         #                  raw (non-truncated) token-level IS weights      (ref bounds: 0.999–1.002)
         #   Blog: https://yingru.notion.site/When-Speed-Kills-Stability-Demystifying-RL-Collapse-from-the-Training-Inference-Mismatch-271211a558b7808d8b12d403fd15edda
+        tis_clipped_fraction = 0.0
         if self.truncated_importance_sampling_ratio is not None:
+            with torch.no_grad():
+                # Divide by global_valid_toks (not local mask.sum()) so that np.sum aggregation in grpo.py gives the correct global fraction
+                tis_clipped_fraction = (
```
Comment (yfw) [inline nemo_rl/algorithms/loss/loss_functions.py]: [`loss_functions.py:460-470`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/nemo_rl/algorithms/loss/loss_functions.py#L460)

`tis_clipped_fraction` is computed *outside* the `truncated_importance_sampling_type` branch, so it runs for **all** IS types, but it only counts tokens with weight above the upper ratio.

- For `tis`: this equals the existing `is_oob_ratio` (`1 - mean(weight <= ratio)`) → redundant.
- For `icepop`: the filter is two-sided (it also zeros tokens below the lower bound), so an upper-bound-only count is incomplete.
- For `seq-mask-tis`: masking is **sequence-level** on the geomean IS ratio, so a token-level upper-bound count doesn't reflect what's masked → misleading under the "clipped" name.

Suggest gating it to `truncated_importance_sampling_type == "tis"` or unifying with the per-type `is_oob_ratio` as a single metric. (Matches @yuki-97's thread above.)

File: tests/unit/environments/test_nemo_gym.py [hunk 6]
```diff
@@ -47,6 +52,48 @@ def test_nemo_gym_stub_module():
     )
 
 
+@pytest.mark.nemo_gym
```
Comment (yfw) [inline tests/unit/environments/test_nemo_gym.py]: [`test_nemo_gym.py:55`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/tests/unit/environments/test_nemo_gym.py#L55)

`_detect_invalid_tool_call_and_malformed_thinking` is a pure-Python function, but this test is marked `@pytest.mark.nemo_gym` (excluded from the default L0 suite) **and** lives in a file whose module-level [`from nemo_rl.models.generation.vllm import VllmGeneration`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/tests/unit/environments/test_nemo_gym.py#L34) gates collection on the vllm extra. Net: this fast unit test only runs in the heaviest vllm+nemo_gym suite and gets near-zero CI coverage. Consider moving the detector + its test to a vllm-free module (e.g. `test_nemo_gym_utils.py`) and dropping the mark so it runs in default L0.

File: nemo_rl/algorithms/grpo.py (hunk 3 above)
Comment (yfw) [inline nemo_rl/algorithms/grpo.py:1119]: [`grpo.py:1083`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/nemo_rl/algorithms/grpo.py#L1083)

`_apply_message_level_advantage_penalties` is missing a docstring. Suggested:

```python
    """Overwrite advantages for flagged assistant-message token spans.

    For each assistant message flagged by the NeMo-Gym detector as an invalid
    tool call or malformed thinking, overwrite that message's advantage span in
    ``train_data["advantages"]`` with the configured negative value. No-op when
    neither ``grpo.invalid_tool_call_advantage`` nor
    ``grpo.malformed_thinking_advantage`` is set.

    Args:
        train_data: Training batch; ``advantages`` is modified in place.
        repeated_batch: Batch carrying ``message_log`` with per-message flags.
        master_config: Master config holding the grpo penalty values.
        log_config: If True, print the configured penalty values once.
    """
```

File: nemo_rl/algorithms/grpo_sync.py [hunk 7]
```diff
@@ -83,6 +83,29 @@
 from nemo_rl.utils.timer import TimeoutChecker, Timer
 from nemo_rl.utils.venvs import make_actor_runtime_env
 
+
+def _raise_if_message_level_advantage_penalties_enabled(
```
Comment (yfw) [inline nemo_rl/algorithms/grpo_sync.py:88]: [`grpo_sync.py:87`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/nemo_rl/algorithms/grpo_sync.py#L87)

`_raise_if_message_level_advantage_penalties_enabled` is missing a docstring. Suggested:

```python
    """Raise if message-level advantage penalties are set in the sync trainer.

    Message-level advantage penalties are not supported with
    ``data_plane.enabled=true``. Raises NotImplementedError listing the
    offending keys so the user can disable them or switch to the legacy GRPO
    trainer.
    """
```

File: nemo_rl/environments/nemo_gym.py [hunk 8]
```diff
@@ -22,11 +22,74 @@
...
     initial_global_config_dict: Dict[str, Any]
+    invalid_tool_call_patterns: NotRequired[
+        List[str] | None
+    ]  # Substrings in assistant text content that indicate an invalid tool call
+    thinking_tags: NotRequired[
+        List[str] | None
+    ]  # Thinking tags to check for malformed usage
+
+
+def _detect_invalid_tool_call_and_malformed_thinking(
```
Comment (yfw) [inline nemo_rl/environments/nemo_gym.py:56]: [`nemo_gym.py:46`](https://github.com/NVIDIA-NeMo/RL/blob/a31cf6b4f0280ddc7a9836dc212dac0a339c5af7/nemo_rl/environments/nemo_gym.py#L46)

`_detect_invalid_tool_call_and_malformed_thinking` is missing a docstring — worth documenting the intentional asymmetry (a *content* message flags any thinking tag, a *reasoning* summary flags only repeated tags). Suggested:

```python
    """Flag a NeMo-Gym output item as an invalid tool call / malformed thinking.

    Inspects the final output item of a model turn. For a final *content*
    message, any thinking tag is malformed (thinking should never leak into the
    answer); for a *reasoning* summary, only a repeated tag (count > 1) is
    malformed (a single pair is expected). A textual tool-call pattern in either
    indicates an invalid (unexecuted) tool call.

    Returns:
        (is_invalid_tool_call, has_malformed_thinking).
    """
```

File: nemo_rl/algorithms/loss/loss_functions.py (hunk 5 above)
Comment (macandro96) [reply nemo_rl/algorithms/loss/loss_functions.py]: addressed - can you check if it needs changes?

File: nemo_rl/algorithms/loss/loss_functions.py (hunk 1 above)
Comment (macandro96) [reply nemo_rl/algorithms/loss/loss_functions.py]: addressed - can you check if it needs changes?

File: nemo_rl/algorithms/grpo.py (hunk 3 above)
Comment (macandro96) [reply nemo_rl/algorithms/grpo.py:1119]: `advantage_estimator.py` seem to contain different advantage estimator class. Semantically it would be better to put it in a new `advantage_estimator_helper.py` or reuse the `utils.py`. For now I've kept it in `grpo.py`. Maybe when we implement it inside `grpo_sync.py` too, we can move it elsewhere?

**Review by yuki-97 — DISMISSED** [review]
LGTM but looks there's a test failed, could you help to fix?

**Review by yuki-97 — DISMISSED** [review]

**Review by yfw — APPROVED** [review]
